base/tests/utils/integration/com/google/idea/blaze/base/ExternalWorkspaceFixture.java

```diff
@@ -0,0 +1,77 @@
+package com.google.idea.blaze.base;
+
+import com.google.idea.blaze.base.ideinfo.ProjectDataInterner;
+import com.google.idea.blaze.base.lang.buildfile.psi.BuildFile;
+import com.google.idea.blaze.base.lang.buildfile.references.LabelUtils;
+import com.google.idea.blaze.base.model.BlazeProjectData;
+import com.google.idea.blaze.base.model.primitives.ExternalWorkspace;
+import com.google.idea.blaze.base.model.primitives.Label;
+import com.google.idea.blaze.base.model.primitives.TargetName;
+import com.google.idea.blaze.base.model.primitives.WorkspacePath;
+import com.google.idea.blaze.base.model.primitives.WorkspaceRoot;
+import com.google.idea.blaze.base.sync.data.BlazeProjectDataManager;
+import com.intellij.psi.PsiFile;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.idea.blaze.base.settings.ui.ProjectViewUi.getProject;
+import static org.junit.Assert.assertNotNull;
+
+public class ExternalWorkspaceFixture {
+  public final ExternalWorkspace w;
+
+  final TestFileSystem fileSystem;
+  WorkspaceFileSystem workspaceFileSystem;
+
+  public ExternalWorkspaceFixture(ExternalWorkspace w, TestFileSystem fileSystem) {
+    this.w = w;
+    this.fileSystem = fileSystem;
+  }
+
+  public WorkspaceRoot getWorkspaceRoot() {
+    BlazeProjectData blazeProjectData = BlazeProjectDataManager.getInstance(getProject()).getBlazeProjectData();
+    assertThat(blazeProjectData).isNotNull();
+
+    File outputBase = blazeProjectData.getBlazeInfo().getOutputBase();
+
+    Path workspaceRootPath = Paths.get(
+        blazeProjectData.getBlazeInfo().getOutputBase().getAbsolutePath(),
+        "external", w.name());
+
+    return new WorkspaceRoot(workspaceRootPath.normalize().toFile());
+  }
+
+  public BuildFile createBuildFile(WorkspacePath workspacePath, String... contentLines) {
+    PsiFile file = getWorkspaceFileSystem().createPsiFile(workspacePath, contentLines);
+    assertThat(file).isInstanceOf(BuildFile.class);
+    return (BuildFile) file;
+  }
+
+  WorkspaceFileSystem getWorkspaceFileSystem() {
+    if (workspaceFileSystem == null) {
```

Comment:
Why do we need to handle null here? All the calls provide file systems, don't they?

Reply:
I'm building a particular filesystem for this workspace root (basically a rooted fs in the `../externa/name`) on demand since it needs access to the test fs which is usually not initialized properly until the Before handlers are run.

Reply:
ok, let's keep it then